### PR TITLE
Declarative clusters and overrides

### LINKED ISSUE #416: cluster overrides are not declarative
The "cluster overrides" feature introduced in v0.10 creates files in git and is not declarative because
- Creating the override requires the `arlon` CLI/API
- Deleting the override (and the cluster) also requires `arlon` CLI/API
This is inconsistent with the fully declarative gen2 clusters (with no overrides) and AppProfiles.
One consequence of this flaw is the overrides feature cannot be added to the `examples/declarative` example.

Aha! Link: https://pf9.aha.io/features/ARLON-376

### PR DESCRIPTION
**Pull request by bcle**
Fixes #416 
Design: https://github.com/arlonproj/arlon/blob/private/leb/416/docs/declarative_clusters.md

### CODE REVIEW

File: api/v1/cluster_types.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+// ClusterSpec defines the desired state of Cluster
+type ClusterSpec struct {
+	// The git location of the cluster template
+	ClusterTemplate RepoSpec `json:"clusterTemplate"`
+	// Optional override specification
+	Override *OverrideSpec `json:"override,omitempty"`
+	// Optional autoscaler specification
+	Autoscaler *AutoscalerSpec `json:"autoscaler,omitempty"`
+	// Optional Arlon Helm chart specification if defaults are not desired
+	ArlonHelmChart *RepoSpec `json:"arlonHelmChart,omitempty"`
```
Comment (Rohitrajak1807) [inline api/v1/cluster_types.go:35]: This refers to the manifests in `pkg/cluster/manifest` correct?

Comment (bcle) [reply api/v1/cluster_types.go:35]: Yes

**Review by Rohitrajak1807 — CHANGES_REQUESTED** [review]
Minor suggestions, but otherwise it's all great, and we should also remove the commented code

File: cmd/cluster/create.go [hunk 2]
```diff
@@ -69,8 +69,12 @@ func createClusterCommand() *cobra.Command {
 				if err == nil {
 					return fmt.Errorf("arlon cluster already exists")
 				}
+				patchContent, err := os.ReadFile(overridesPath)
+				if err == nil {
+					return fmt.Errorf("failed to read patch file: %s", err)
+				}
```
[the remark this answers is not in the document]
Comment (bcle) [reply cmd/cluster/create.go:75]: fixed

File: controllers/cluster_controller.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+			return r.UpdateState(ctx, log, &cr, "retrying", msg, retryDelayAsResult)
+		}
+		return r.UpdateState(ctx, log, &cr, "created",
+			"cluster app creation successful", ctrl.Result{})
+	}
+
+	// Sync profile annotation from Cluster to cluster app if necessary
+	if cr.Annotations != nil && (clusterApp.Annotations == nil ||
+		cr.Annotations[arlonapp.ProfilesAnnotationKey] !=
+			clusterApp.Annotations[arlonapp.ProfilesAnnotationKey]) {
```
[the remark this answers is not in the document]
Comment (bcle) [reply controllers/cluster_controller.go:228]: It actually got more complicated: there is a case I forgot, when the source annotations don't exist (nil), but the destination is not nil. This required another test followed by custom logic. In the end, it's actually more complicated if we refactor the 2 tests into separate functions. See updated code.

[the remark this answers is not in the document]
Comment (Rohitrajak1807) [reply controllers/cluster_controller.go:228]: This bit `clusterApp.Annotations == nil` correct?

[the remark this answers is not in the document]
Comment (Rohitrajak1807) [reply controllers/cluster_controller.go:228]: That is fine then

**Review by Rohitrajak1807 — APPROVED** [review]
Looks good.
There is some commented code in `pkg/gitutils/manifests.go` but we can take care of later.

**Review by cruizen — APPROVED** [review]

**Comment by bcle** [comment]
Review feedback addressed in last commit. Also fixed the profile sync case where the Cluster no longer has annotations, but the Application resource still does.

**Pull request merged**